frontends/p4/unusedDeclarations.h

```diff
@@ -22,10 +22,38 @@ limitations under the License.
 
 namespace P4 {
 
+/** @brief Removes unused declarations.
+ *
+ * The following kinds of nodes are not removed even if they are unreferenced:
+ *  - IR::Declaration_MatchKind
+ *  - IR::Parameter
+ *  - IR::Type_Error
+ *  - IR::Type_Extern
+ *  - IR::Type_Method
+ *  - IR::Type_StructLike
+ *
+ * If @warned is non-null, unused IR::P4Table and IR::Declaration_Instance
+ * nodes are stored in @warned if they are unused and removed by this pass.
```

Comment:
In addition, a warning message is emitted only if the declaration was not previously in `warned` -- that is in fact the main point of `warned` -- to suppress duplicate warnings